mesonbuild/scripts/coverage.py

```diff
@@ -14,87 +14,135 @@
+                                   '-e', subproject_root,
+                                   '-o', os.path.join(log_dir, 'coverage.xml'),
+                                   ])
+            outfiles.append(('Xml', pathlib.Path(log_dir, 'coverage.xml')))
+        elif outputs:
+            print('gcovr needed to generate Xml coverage report')
+            exitcode = 1
+
+    if not outputs or 'text' in outputs:
+        if gcovr_exe:
+            subprocess.check_call([gcovr_exe,
+                                   '-r', gcovr_rootdir,
+                                   '-e', subproject_root,
+                                   '-o', os.path.join(log_dir, 'coverage.txt'),
+                                   ])
+            outfiles.append(('Text', pathlib.Path(log_dir, 'coverage.txt')))
+        elif outputs:
+            print('gcovr needed to generate text coverage report')
+            exitcode = 1
+
+    if not outputs or 'html' in outputs:
+        if lcov_exe and genhtml_exe:
+            htmloutdir = os.path.join(log_dir, 'coveragereport')
+            covinfo = os.path.join(log_dir, 'coverage.info')
+            initial_tracefile = covinfo + '.initial'
+            run_tracefile = covinfo + '.run'
+            raw_tracefile = covinfo + '.raw'
+            subprocess.check_call([lcov_exe,
+                                   '--directory', build_root,
+                                   '--capture',
+                                   '--initial',
+                                   '--output-file',
+                                   initial_tracefile])
+            subprocess.check_call([lcov_exe,
+                                   '--directory', build_root,
+                                   '--capture',
+                                   '--output-file', run_tracefile,
+                                   '--no-checksum',
+                                   '--rc', 'lcov_branch_coverage=1',
+                                   ])
+            # Join initial and test results.
+            subprocess.check_call([lcov_exe,
+                                   '-a', initial_tracefile,
+                                   '-a', run_tracefile,
+                                   '-o', raw_tracefile])
+            # Remove all directories outside the source_root from the covinfo
+            subprocess.check_call([lcov_exe,
+                                   '--extract', raw_tracefile,
+                                   os.path.join(source_root, '*'),
+                                   '--output-file', covinfo])
+            # Remove all directories inside subproject dir
+            subprocess.check_call([lcov_exe,
+                                   '--remove', covinfo,
+                                   os.path.join(subproject_root, '*'),
+                                   '--output-file', covinfo])
+            subprocess.check_call([genhtml_exe,
+                                   '--prefix', build_root,
```

Comment:
Actually I found one bug. But this was already in the previous version, so not your fault! This should be `'--prefix', source_root,`.

The `--prefix` flag strips the root part from from the source filenames so that the report is easier to read.

Reply:
Both the previous ninja coverage-html target and meson --internal coverage currently uses --prefix build_root.
I don't disagree with the change but I think making a separate issue for it would be cleaner. Only objection I can think of is that it makes the report depending on the output directory name if you have any generated sources (that now will be <builddir>/sourcefile instead of just sourcefile).

Reply:
Yeah let's do that in another PR. I'll open an issue

Reply:
Created #3260